### ISSUE TITLE
Improve DNS lookup performance by working around ndots

### ISSUE THREAD

**Issue by lfrancke** [stackabletech/issues#656]
The default `resolv.conf` file generated by a kubelet contains an option called `ndots: 5` which can result in 4 extra DNS requests for every lookup which we can prevent.

Example `resolv.conf`:
```
search default.svc.cluster.local svc.cluster.local cluster.local localdomain
nameserver 10.96.0.10
options ndots:5
```

Example DNS request for `foo-service.foo.svc.cluster.local` from within a Pod results in these DNS requests in Core DNS: 
```
"A IN foo-service.foo.svc.cluster.local.default.svc.cluster.local. udp 100 false 4096" NXDOMAIN qr,aa,rd 170 0.000081963s
"A IN foo-service.foo.svc.cluster.local.svc.cluster.local. udp 92 false 4096" NXDOMAIN qr,aa,rd 162 0.000039183s
"A IN foo-service.foo.svc.cluster.local.cluster.local. udp 88 false 4096" NXDOMAIN qr,aa,rd 158 0.000030127s
"A IN foo-service.foo.svc.cluster.local.localdomain. udp 86 false 4096" NXDOMAIN qr,rd,ra 63 0.000683273s
"A IN foo-service.foo.svc.cluster.local. udp 74 false 4096" NOERROR qr,aa,rd 100 0.000030406s
```

The documentation for `ndots` says:
```
sets a threshold for the number of dots which must appear in a name before an initial absolute query will be made. The default for n is 1, meaning that if there are any dots in a name, the name will be tried first as an absolute name before any search list elements are appended to it.
```

So, it tries each of the entries from the `search` list in order before it tries an absolute query.
In our case we try to use FQDNs everywhere to avoid ambiguity which means these extra lookups are not necessary but you can see that even using this FQDN does not help.

I see two options to work around this:

## Option 1: Add "."

If we appended a "." to all our lookups we would have 5 dots and an absolute lookup will be performed. A lookup for `foo-service.foo.svc.cluster.local.` results in:
```
"A IN foo-service.foo.svc.cluster.local. udp 74 false 4096" NOERROR qr,aa,rd 100 0.000051045s
```

## Option 2: Change DNS options for our Pods

By creating all our Pods with this additional config we can work around the problem "globally":

```
spec:
  dnsConfig:
    options:
      - name: ndots
        value: "1"
```
https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config

## Option 3: Do both

Option 1 might be complicated to catch all cases if anyone uses pod overrides or if we assemble a name somewhere in a non-standard location.
Option 2 might - in theory - be overridden.
We could just apply both. Everything still works without this as it's just an optimization and I don't see a downside in just applying both options.

**Comment by nightkr** [comment]
I definitely think we should do at least 1. I see the argument for 2, but would break user configuration that relies on it (either accidentally, or as a workaround to keep *their* environments isolated).

I'd rather start moving towards maybe eventually seeing the default changed in a hypothetical Pod/v2 API.

**Comment by lfrancke** [comment]
Can you elaborate? Can you give me an example in which user scenario this would break?

**Comment by eminaktas** [comment]
I would suggest ndots as 2 which could be the best option. Please see my suggestion for dns configuration [here](https://github.com/eminaktas/kubedns-shepherd/blob/df0c3bf004f51aa9b82cc10e363c69bd51321539/chart/kubedns-shepherd/values.yaml#L65-L88)

If you go with 1, the DNS queries with 1 or more dots is treated as an FQDN. For example, `dummy-service.another-namespace` will fail because it won't complete the query with `svc.cluster.local`. You would simply block the app to reach apps in another namespace.

I built [kubedns-shepherd ](https://github.com/eminaktas/kubedns-shepherd) to have more control over DNS settings and even customizing it at the host level to optimize more. For example, [this ](https://github.com/kubernetes-sigs/kubespray/blob/9d39d35dfbe51ff4e9d9beac77afb0dc0b454685/inventory/sample/group_vars/k8s_cluster/k8s-cluster.yml#L162 ) configuration in Kubespray repository.

**Comment by lfrancke** [comment]
Hey @eminaktas, thank you for the hints. That's a good point!
We haven't made any decisions here yet so this is coming in at a good time.

As an aside: I did look through the README for `kubedns-shepherd` and saw your snippet on auto-discovery of `clusterDomain` etc.
I am not entirely sure how you stumbled over this issue but you might have seen it pop up on #sig-network in Slack? If not, then this blog post might be of interest to you: https://stackable.tech/en/kubernetes-clusterdomain-setting/ and potentially this follow-up issue: https://github.com/stackabletech/issues/issues/662

**Comment by eminaktas** [comment]
Thanks @lfrancke.

I stumbled on this issue thanks to Antonio Ojea. He shared on X. I really enjoyed the article and I have also been looking for an API to get clusterDomain along with clusterDNS. I will definitely keep eye on this topic.

Can we also include clusterDNS? I am not sure if it is already exists via an API.

I am able to get the information like you did in the article.

```
kubectl get --raw /api/v1/nodes/minikube/proxy/configz | jq .kubeletconfig.clusterDNS
```

**Comment by maltesander** [comment]
Im with Nat, i think Option 1 will instantly reduce DNS queries in "default" setups significantly, while not having any other real drawbacks.
Since we will always work with FQDNs, this is the only "constant" we can count on here.

Im also fine with simply documenting this from a Stackable point of view. E.g. "operators always write FQDNs ... in order to decrease DNS queries set ndots to 4" (assuming we do not add another (5th) "." in the end).

Concerning Option 2 or setting/overriding `ndots`, I also do not see the point to "doctor" around customer specific settings. I dont think that 1, 2 or 5 are "correct" or "best". This is simply too undeterministic, so im all for doing Option 1:
- In code (adding another "." to FQDNs) -> needs some investigation, IMHO some operators never use or write FQDNs (Airflow?).
- In docs (recommend ndots setting to reduce DNS queries for the overall SDP)

**Comment by lfrancke** [comment]
Thank you.

I think "just" adding a "." to the discovered or configured `clusterDomain` should be enough to catch most cases, no? It's okay to only have a 40-60% solution here. Anything is better than nothing.

**Comment by maltesander** [comment]
Yes, if we assume that 40-60% have "default" ndots 5. I think doing less is more here, since it just depends too heavily on customers, cluster setups etc., having this properly documented should be the first step, going for option 1 afterwards is not a huge change either.

**Comment by lfrancke** [comment]
@eminaktas You might be interested in https://github.com/kubernetes/enhancements/pull/4972 and adding your feedback there

**Comment by dervoeti** [comment]
Fixed by https://github.com/stackabletech/operator-rs/pull/939

Summary:
We changed the default domain from `cluster.local` to `cluster.local.` (we appended a dot) to make it a FQDN, which reduces DNS queries. We found that with a trailing dot the `search` domains are never used, no matter which value `ndots` is set to (might depend on the client, but I was not able to find a client that treats this different).
If the user specifies a custom cluster domain, we don't modify it in any way, but we print a warning to the log if it does not end with a dot. The warning informs the user about the potential improvement and refers to this issue.

**Comment by dervoeti** [comment]
Update:
We decided not to make `cluster.local.` the default cluster domain and instead use the previous default (`cluster.local`, without a trailing dot).

Reasoning: We came across some issues with TLS certificates and Kerberos keytabs and decided to make this behavior (FQDN domains for services/pods) opt-in for now. The issues are likely solvable but may require manual migration effort. We still wanted to keep experimental support for it, but only if a user opts in by explicitly setting `KUBERNETES_CLUSTER_DOMAIN` to a domain with a trailing dot.

PR with this change: https://github.com/stackabletech/operator-rs/pull/947

**Comment by dervoeti** [comment]
Update:
While FQDN cluster domains are already supported experimentally in the current dev versions, we experienced some issues related to mTLS. Once https://github.com/stackabletech/secret-operator/pull/564 is merged, setting `KUBERNETES_CLUSTER_DOMAIN` to a FQDN should improve DNS performance overall and most setups should be supported.

**Comment by dervoeti** [comment]
mTLS issues are resolved, we now support this experimentally. We're currently not aware of any issues but don't have enough experience with this feature, and since we experienced some unexpected side effects during testing that had to be fixed we don't consider it stable as of now. Starting with SDP 25.3, which will be released soon, users can opt in to this improvement by specifying a FQDN (with a trailing dot) as Kubernetes cluster domain for our operators. This will also be mentioned in [our docs](https://docs.stackable.tech/home/stable/guides/kubernetes-cluster-domain/) with the release of 25.3.

**Comment by dervoeti** [comment]
## Release Notes

* Users can now opt-in to experimental support for improved DNS performance by specifying a fully qualified domain name (with a trailing dot) as Kubernetes cluster domain for our operators. This can reduce the amount of DNS queries within the cluster, but might have side effects, hence we consider the support experimental for now. Take a look at [our documentation](https://docs.stackable.tech/home/stable/guides/kubernetes-cluster-domain/) to find out how to enable this improvement.

**Comment by lfrancke** [comment]
Should this be marked as experimental in the docs?

**Comment by dervoeti** [comment]
> Should this be marked as experimental in the docs?

I mention it in the docs, the link in my comment goes to "stable" and currently "stable" is still pointing at 24.11. These are the upcoming docs: https://docs.stackable.tech/home/nightly/guides/kubernetes-cluster-domain/
